bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/datainteg/DataIntegrityCheckImpl.java

```diff
@@ -0,0 +1,554 @@
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.bookie.BookieException;
+import org.apache.bookkeeper.bookie.LedgerStorage;
+import org.apache.bookkeeper.bookie.LedgerStorage.StorageState;
+import org.apache.bookkeeper.client.BKException;
+import org.apache.bookkeeper.client.BookKeeperAdmin;
+import org.apache.bookkeeper.client.api.LedgerMetadata;
+import org.apache.bookkeeper.common.concurrent.FutureUtils;
+import org.apache.bookkeeper.meta.LedgerManager;
+import org.apache.bookkeeper.net.BookieId;
+
+/**
+ * An implementation of the DataIntegrityCheck interface.
+ */
+@Slf4j
+public class DataIntegrityCheckImpl implements DataIntegrityCheck {
+    private static final int MAX_INFLIGHT = 300;
+    private static final int MAX_ENTRIES_INFLIGHT = 3000;
+    private static final int ZK_TIMEOUT_S = 30;
+    private final BookieId bookieId;
+    private final LedgerManager ledgerManager;
+    private final LedgerStorage ledgerStorage;
+    private final EntryCopier entryCopier;
+    private final BookKeeperAdmin admin;
+    private final Scheduler scheduler;
+    private final AtomicReference<Map<Long, LedgerMetadata>> ledgersCacheRef =
+        new AtomicReference<>(null);
+    private CompletableFuture<Void> preBootFuture;
+
+    public DataIntegrityCheckImpl(BookieId bookieId,
+                                  LedgerManager ledgerManager,
+                                  LedgerStorage ledgerStorage,
+                                  EntryCopier entryCopier,
+                                  BookKeeperAdmin admin,
+                                  Scheduler scheduler) {
+        this.bookieId = bookieId;
+        this.ledgerManager = ledgerManager;
+        this.ledgerStorage = ledgerStorage;
+        this.entryCopier = entryCopier;
+        this.admin = admin;
+        this.scheduler = scheduler;
+    }
+
+    @Override
+    public CompletableFuture<Void> runPreBoot(String reason) {
+        // we only run this once, it could be kicked off by different checks
+        synchronized (this) {
+            if (preBootFuture == null) {
+                preBootFuture = runPreBootSequence(reason);
+            }
+        }
+        return preBootFuture;
+
+    }
+
+    private CompletableFuture<Void> runPreBootSequence(String reason) {
+        String runId = UUID.randomUUID().toString();
+        log.info("Event: {}, RunId: {}, Reason: {}", Events.PREBOOT_START, runId, reason);
+        try {
+            this.ledgerStorage.setStorageStateFlag(StorageState.NEEDS_INTEGRITY_CHECK);
+        } catch (IOException ioe) {
+            log.error("Event: {}, RunId: {}", Events.PREBOOT_ERROR, runId, ioe);
+            return FutureUtils.exception(ioe);
+        }
+
+        MetadataAsyncIterator iter = new MetadataAsyncIterator(scheduler,
+                ledgerManager, MAX_INFLIGHT, ZK_TIMEOUT_S, TimeUnit.SECONDS);
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        Map<Long, LedgerMetadata> ledgersCache =
```

Comment:
should we be concerned about memory requirements here in case of large disks (to many ledgers/metadata with many segments)? See https://github.com/apache/bookkeeper/pull/1949